### PR TITLE
[CELEBORN-243][CELEBORN-245][IMPROVEMENT] Create push client failed and connection failed cause push failed should have their own ERROR type

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Push data connect failed is a critical error, we should have a divided error status for these case and add it to blacklist  and reserve for some time.
In this PR we add four more status code for push fail status:

1.    PUSH_DATA_CONNECT_FAIL_MASTER: Throw only when shuffle client create  connection to master worker failed 
2.   PUSH_DATA_CONNECT_FAIL_SLAVE: : Throw only when master worker create  connection to slave worker failed 
3.   PUSH_DATA_CONNECTION_FAIL_MASTER:  Throw only when shuffle client create  connection to master worker success but push data throw connection exception 
4.   PUSH_DATA_CONNECTION_FAIL_SLAVE:  Throw only when master worker create  connection to slave worker success but push data throw connection exception

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture Have change the code as discussed this morning, pls take a look

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 1]
```diff
@@ -1516,7 +1561,13 @@ public void onFailure(Throwable e) {
...
     } catch (Exception e) {
       logger.warn("PushData byteBuf failed", e);
-      callback.onFailure(new Exception(getPushDataFailCause(e.getMessage()).toString(), e));
+      callback.onFailure(
+          new Exception(
+              StatusCode.PUSH_DATA_CONNECT_FAIL_MASTER.getMessage()
+                  + "! "
+                  + e.getMessage()
+                  + ". "
+                  + location.toString()));
```
Comment (waitinfuture) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:1570]: ditto

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 2]
```diff
@@ -268,8 +275,9 @@ class PushDataHandler extends BaseMessageHandler with Logging {
             peer.getReplicatePort)
           if (unavailablePeers.containsKey(peerWorker)) {
             pushData.body().release()
-            wrappedCallback.onFailure(
-              new Exception(s"Peer $peerWorker unavailable for $location!"))
+            callback.onFailure(
```
Comment (waitinfuture) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:278]: wrappedCallback.onFailure will call ```workerSource.incCounter(WorkerSource.PushDataFailCount)```. If we change to callback here, we should also call incCounter here

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 3]
```diff
@@ -285,8 +293,10 @@ class PushDataHandler extends BaseMessageHandler with Logging {
             case e: Exception =>
               pushData.body().release()
               unavailablePeers.put(peerWorker, System.currentTimeMillis())
-              wrappedCallback.onFailure(
-                new Exception(s"Push data to peer $peerWorker failed for $location", e))
+              callback.onFailure(
```
Comment (waitinfuture) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:296]: ditto

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 4]
```diff
@@ -447,7 +465,8 @@ class PushDataHandler extends BaseMessageHandler with Logging {
             peer.getReplicatePort)
           if (unavailablePeers.containsKey(peerWorker)) {
             pushMergedData.body().release()
-            wrappedCallback.onFailure(new Exception(s"Peer $peerWorker unavailable for $location!"))
+            callback.onFailure(new Exception(
```
Comment (waitinfuture) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:468]: ditto

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 5]
```diff
@@ -466,9 +485,10 @@ class PushDataHandler extends BaseMessageHandler with Logging {
             case e: Exception =>
               pushMergedData.body().release()
               unavailablePeers.put(peerWorker, System.currentTimeMillis())
-              wrappedCallback.onFailure(new Exception(
-                s"Push data to peer $peerWorker failed for $location",
-                e))
+              callback.onFailure(
```
Comment (waitinfuture) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:488]: ditto

File: common/src/main/java/org/apache/celeborn/common/protocol/message/StatusCode.java [hunk 6]
```diff
@@ -68,7 +68,11 @@ public enum StatusCode {
   REGION_FINISH_FAIL_SLAVE(36),
   REGION_FINISH_FAIL_MASTER(37),
 
-  PUSH_DATA_TIMEOUT(38);
+  PUSH_DATA_CONNECT_FAIL_MASTER(38),
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/protocol/message/StatusCode.java:71]: ```PUSH_DATA_CONNECT_FAIL_``` and ```PUSH_DATA_CONNECTION_FAIL_``` is quite confusing. IMO we can rename to ```PUSH_DATA_CREATE_CONNECTION_FAIL_``` and ```PUSH_DATA_CONNECTION_EXCEPTION_```

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 7]
```diff
@@ -739,12 +739,22 @@ public void onSuccess(ByteBuffer response) {
...
                 return;
               }
 
               if (remainReviveTimes <= 0) {
-                callback.onFailure(e);
+                callback.onFailure(
+                    new Exception(
+                        cause.getMessage()
+                            + "! Push data to master worker of "
+                            + loc.toString()
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:753]: Done

File: common/src/main/java/org/apache/celeborn/common/protocol/message/StatusCode.java (hunk 6 above)
Comment (AngersZhuuuu) [reply common/src/main/java/org/apache/celeborn/common/protocol/message/StatusCode.java:71]: Done

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala:279]: > wrappedCallback.onFailure will call `workerSource.incCounter(WorkerSource.PushDataFailCount)`. If we change to callback here, we should also call incCounter here

Done

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**